UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,612 @@
+  EfiMemoryTop. And EfiFreeMemoryBottom and EfiFreeMemoryTop should
+  be inside the HOB region.
+
+  @param[in] EfiMemoryBottom       Total memory start address
+  @param[in] EfiMemoryTop          Total memory end address.
+  @param[in] EfiFreeMemoryBottom   Free memory start address
+  @param[in] EfiFreeMemoryTop      Free memory end address.
+
+  @return   The pointer to the handoff HOB table.
+
+**/
+EFI_HOB_HANDOFF_INFO_TABLE *
+EFIAPI
+HobConstructor (
+  IN VOID  *EfiMemoryBottom,
+  IN VOID  *EfiMemoryTop,
+  IN VOID  *EfiFreeMemoryBottom,
+  IN VOID  *EfiFreeMemoryTop
+  );
+
+/**
+  It will parse FDT based on DTB from bootloaders.
+
+  @param[in]  FdtBase               Address of the Fdt data.
+
+  @retval EFI_SUCCESS               If it completed successfully.
+  @retval Others                    If it failed to parse DTB.
+**/
+UINTN
+ParseDtb (
+  IN VOID  *FdtBase
+  )
+{
+  VOID                                *Fdt;
+  INT32                               Node;
+  INT32                               Property;
+  INT32                               Depth;
+  FDT_NODE_HEADER                     *NodePtr;
+  CONST FDT_PROPERTY                  *PropertyPtr;
+  CONST CHAR8                         *TempStr;
+  INT32                               TempLen;
+  UINT32                              *Data32;
+  UINT64                              *Data64;
+  UINT64                              StartAddress;
+  INT32                               SubNode;
+  UINT64                              NumberOfBytes;
+  UINT32                              Attribute;
+  UNIVERSAL_PAYLOAD_SERIAL_PORT_INFO  *Serial;
+  //  UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES    *PciRootBridgeInfo;
+  UEFI_PAYLOAD_DEBUG_PRINT_ERROR_LEVEL  *DebugPrintErrorLevelInfo;
+  UNIVERSAL_PAYLOAD_BASE                *PayloadBase;
+  UNIVERSAL_PAYLOAD_ACPI_TABLE          *PldAcpiTable;
+  UNIVERSAL_PAYLOAD_SMBIOS_TABLE        *SmbiosTable;
+  EFI_PEI_GRAPHICS_INFO_HOB             *GraphicsInfo;
+  EFI_PEI_GRAPHICS_DEVICE_INFO_HOB      *GraphicsDev;
+  UINT8                                 SizeOfMemorySpace;
+  UINT64                                FrameBufferBase;
+  UINT64                                FrameBufferSize;
+
+  Fdt             = FdtBase;
+  Depth           = 0;
+  FrameBufferBase = 0;
+  FrameBufferSize = 0;
+
+  DEBUG ((DEBUG_INFO, "FDT = 0x%x  %x\n", Fdt, Fdt32ToCpu (*((UINT32 *)Fdt))));
+  DEBUG ((DEBUG_INFO, "Start parsing DTB data\n"));
+
+  //
+  // Build SMBIOS HOB .
+  //
+  SmbiosTable = BuildGuidHob (&gUniversalPayloadSmbios3TableGuid, sizeof (UNIVERSAL_PAYLOAD_SMBIOS_TABLE) + sizeof (SMBIOS_TABLE_3_0_ENTRY_POINT));
+  ASSERT (SmbiosTable != NULL);
+  if (SmbiosTable == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  SmbiosTable->Header.Revision = UNIVERSAL_PAYLOAD_SMBIOS_TABLE_REVISION;
+  SmbiosTable->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_SMBIOS_TABLE);
+
+  PropertyPtr = FdtGetProperty (Fdt, 0, "smbios", &TempLen);

```

Comment:
What happens when there is no smbios property and it returns NULL? This applies to ther nodes as well. Have you done -ve testing on this code?